packages/flutter_tools/lib/src/test/coverage_collector.dart

```diff
@@ -12,39 +12,70 @@ import '../base/io.dart';
 import '../dart/package_map.dart';
 import '../globals.dart';
 
+/// Class that represents a pending task of coverage data collection.
+/// Instances of this class are obtained when a process starts running code
+/// by calling [CoverageCollector.addTask]. Then, when the code has run to
+/// completion (all the coverage data has been recorded), the task is started
+/// to actually collect the coverage data.
+abstract class CoverageCollectionTask {
+  /// Starts the task of collecting coverage. Returns a future that completes
+  /// when coverage has been collected.
+  ///
+  /// This should be called when the code whose coverage data is being
+  /// collected has been run to completion so that all coverage data has been
+  /// recorded.
+  ///
+  /// A task may only be started once.
+  Future<Null> start();
+
+  /// Indicates whether the task has been started or not.
+  bool get isStarted;
+}
+
+/// Singleton class that's used to collect coverage data during tests.
 class CoverageCollector {
-  static final CoverageCollector instance = new CoverageCollector();
+  static final CoverageCollector instance = new CoverageCollector._();
 
+  CoverageCollector._();
+
+  /// By default, coverage collection is not enabled. Set [enabled] to true
+  /// to turn on coverage collection.
   bool enabled = false;
   int observatoryPort;
 
-  void collectCoverage({
+  /// Adds a coverage collection tasks to the pending queue. The task will not
+  /// begin collecting coverage data until [CoverageCollectionTask.start] is
+  /// called or [finalizeCoverage] is called (which implicitly starts all
+  /// pending tasks).
+  ///
+  /// If this collector is not [enabled], the task will still be added to the
+  /// pending queue. Only when the task is started will the enabled state of
+  /// the collector be consulted.
+  CoverageCollectionTask addTask({
     String host,
     int port,
     Process processToKill,
   }) {
-    if (enabled) {
-      assert(_jobs != null);
-      _jobs.add(_startJob(
-        host: host,
-        port: port,
-        processToKill: processToKill,
-      ));
-    } else {
-      processToKill.kill();
-    }
+    _Task task = new _Task(this, host, port, processToKill);
+    _pendingTasks.add(task);
+    return task;
   }
 
-  Future<Null> _startJob({
-    String host,
-    int port,
-    Process processToKill,
-  }) async {
-    int pid = processToKill.pid;
-    printTrace('collecting coverage data from pid $pid on port $port');
-    Map<String, dynamic> data = await collect(host, port, false, false);
+  Set<_Task> _pendingTasks = new Set<_Task>();
+  List<Future<Null>> _activeTasks = <Future<Null>>[];
+  Map<String, dynamic> _globalHitmap;
+
+  Future<Null> _startTask(_Task task) async {
+    if (!enabled) {
```

Comment:
I would move the bulk of this method into _Task.start(), leaving only the globalHitmap code here. That way the code more clearly separates the Task-related code from the CoverageCollection-related code. (The heuristic I use for deciding this kind of thing is "whose privates does the code deal with". Here you have a lot of `task.foo`.)